apollo-router/src/query_planner/selection.rs

```diff
@@ -48,97 +48,171 @@ pub(crate) struct InlineFragment {
+        );
     }
-    Ok(Some(Value::Object(output)))
+
+    Value::Object(output)
 }
 
-fn select_field<'a>(
-    content: &'a Object,
-    field: &Field,
-    schema: &Schema,
-) -> Result<Option<(&'a ByteString, Value)>, FetchError> {
-    let res = match (
-        content.get_key_value(field.name.as_str()),
-        &field.selections,
-    ) {
-        (Some((k, v)), _) => select_value(v, field, schema).map(|opt| opt.map(|v| (k, v))),
-        (None, _) => Err(FetchError::ExecutionFieldNotFound {
-            field: field.name.to_owned(),
-        }),
+fn is_object_of_type(typename: Option<&str>, condition: &str, schema: &Schema) -> bool {
+    let typename = match typename {
+        None => return false,
+        Some(t) => t,
     };
-    res
-}
 
-fn select_inline_fragment(
-    content: &Object,
-    fragment: &InlineFragment,
-    schema: &Schema,
-) -> Result<Option<Value>, FetchError> {
-    match (&fragment.type_condition, &content.get("__typename")) {
-        (Some(condition), Some(Value::String(typename))) => {
-            if condition == typename || schema.is_subtype(condition, typename.as_str()) {
-                select_object(content, &fragment.selections, schema)
-            } else {
-                Ok(None)
-            }
-        }
-        (None, _) => select_object(content, &fragment.selections, schema),
-        (_, None) => Err(FetchError::ExecutionFieldNotFound {
-            field: "__typename".to_string(),
-        }),
-        (_, _) => Ok(None),
+    if condition == typename {
+        return true;
     }
-}
 
-fn select_value(
-    content: &Value,
-    field: &Field,
-    schema: &Schema,
-) -> Result<Option<Value>, FetchError> {
-    match (content, &field.selections) {
-        (Value::Object(child), Some(selections)) => select_object(child, selections, schema),
-        (Value::Array(elements), Some(_)) => elements
-            .iter()
-            .map(|element| {
-                select_value(element, field, schema)
-                    // if a value cannot be selected from the array element, return Some(Value::Null)
-                    // instead of None, because None will short circuit the iteration and return
-                    // an empty array
-                    .map(|opt| opt.unwrap_or(Value::Null))
-            })
-            .collect::<Result<Vec<Value>, FetchError>>()
-            .map(|v| Some(Value::Array(v))),
-        (value, None) => Ok(Some(value.to_owned())),
-        _ => Ok(None),
+    let object_type = match schema.definitions.types.get(typename) {
+        None => return false,
+        Some(t) => t,
+    };
+
+    let conditional_type = match schema.definitions.types.get(condition) {
+        None => return false,
+        Some(t) => t,
+    };
+
+    if conditional_type.is_interface() || conditional_type.is_union() {
+        return (object_type.is_object() || object_type.is_interface())
+            && schema.is_subtype(condition, typename);
     }
+
+    false
```

Comment:
```suggestion
   (conditional_type.is_interface() || conditional_type.is_union()).then(|| (object_type.is_object() || object_type.is_interface())
            && schema.is_subtype(condition, typename)).unwrap_or(false)
```

not tested, i just wanted to be nitpicky 😂